### DISCUSSION TITLE
Generate dedicated lockfile for individual workspace

### DISCUSSION THREAD

**Discussion by codepunkt** [npm/feedback#650] [CLI] [answered]
I've been racking my brain on this for a while, but couldn't find a solution.

I'm working on a large monorepo with a few dozen packages, using npm workspaces. A lot of these packages are frontend applications and backend services, each of which are deployed as docker containers. To do so, the build process builds a docker image for each of them and has to install the required dependencies in the process. To do so, we copy the root-level `package-lock.json` and the individual `package.json` of that package to the filesystem of the container and perform a npm installation afterwards.

```Dockerfile
COPY package-lock.json .
COPY service/package.json .
RUN npm ci --audit false --fund false

COPY service/ .
RUN npm run build
```

This works fine and will only install the dependencies required for this exact package 👍🏻

However, I have more than three dozen workspaces in the monorepo and update all dependencies in all workspaces automatically. This means the root `package-lock.json` is updated multiple times a day - which leads to my problem.

Due to the `COPY package-lock.json` being one of the, if not the first instruction in all `Dockerfile`, this file changing multiple times a day prevents us from efficiently using [dockers build cache](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#leverage-build-cache). Every time any workspace updates a dependency, all cached docker build layers after this instruction are invalidated and have to be re-executed from scratch for every package.

Both [yarn](https://www.npmjs.com/package/generate-lockfile) (userland) and [pnpm](https://github.com/pnpm/pnpm/tree/main/packages/make-dedicated-lockfile) have tools to generate a specific lockfile for a single workspace out of the hoisted root lockfile.

It should be possible to also do this with npm - but I couldn't find any readymade solutions for it.
What do you think?

**Comment by MylesBorins** [comment, marked as answer]
Hey @codepunkt 

For changes to the CLI the appropriate next step is to open an issue over at [the npm CLI RFC repo](https://github.com/npm/rfcs) to suggest the change.

Thanks for the suggestion!

**Reply by codepunkt** [reply]
Thanks Miles, I opened a RRFC.
https://github.com/npm/rfcs/issues/554

Are the details of the lockfile `v2` format documented somewhere? If they are and someone could point me to them, or if someone could help me understand them in a short call, I could build a prototype in userland.

**Comment by evilj0e** [comment]
It's better to have a separate lock file. The approach of the @codepunkt isn't suitable for me because I have conflict in versions of the dependencies. Some of them are overriden on the package level and I get an error `Invalid: lock file's @types/jest@27.5.0 does not satisfy @types/jest@27.5.1`.
